More matching and sorting options, a couple usability improvements

kvaadit opened the pull request:
I've added a few features that are useful to my own workflow while calling `choose` from scripts. Hence this pull request.

1. _Case-sensitive and Smart-case matching._ I've added a command-line switch (`-C`) that now lets `choose` match rows against the query case-sensitively and also "smartly".

    Specifically, `-C i` is the (default) case-insensitive behavior. `-C I` means case-sensitive matching. And `-C s` means smart-case matching -- ie, case-insensitive if the query is all-lowercase, but case-sensitive if the query contains even 1 uppercase character.

    This feature also respects the `MatchFromBeginning` option (`-z`).

2. _Match whole words instead of characters._ The command-line switch `-W` now asks `choose` to match whole words (ie, contiguous characters) instead of individual characters.

    For example, _without_ this option enabled, the query `se` will match both `Seattle` and `Singapore`. But _with_ this option, it will only match `Seattle`.

    This feature also respects the `MatchFromBeginning` option (`-z`).

3. _`-S` now means "skip scoring/sorting options"._ That is, list matching options in whatever order they appeared in the input.

4. _Auto-select single choice._ I added the command-line switch `-1`, which means: If at any time there's only one choice left (either at the start, or after the user types in a few query characters), select that choice immediately instead of waiting for the user to hit Enter.

5. _Allow first choice to contain leading spaces._ I noticed that when the first choice contains leading spaces, `choose` would strip them out. But it wouldn't do this for any subsequent choice. I fixed this behavior.

6. _Append to initial query instead of replacing it._ When an initial query is provided, I noticed that it's already selected when `choose` pops up. So, whatever the user types subsequently ends up replacing the initial query rather than appending to it. I fixed this.

Commit:
Allow first choice to contain leading whitespace

My use case presents choose with a list of options, where each option is a row
containing a few different columns. These columns can be right aligned, and are
separated by one or more spaces within a row.

This way, each row is just a string, but choose can still present to the user a
set of options where each option contains a bunch of neatly arranged columns.
This works so long as the string representation of each row is properly
formatted -- with the right width and alignment for each column and so forth.

I found that when the first column is right-aligned, it can happen that the
first row starts with a space. But choose will get rid of all such leading
spaces, thereby messing with my neatly arranged columns.

This commit is a one-line fix for this.

Commit:
Add commandline switch: auto-select single choice

After filtering the available choices with the query string, sometimes it may
turn out that only 1 choice remains. In this case, one may want choose to just
select this one choice without waiting for the user to hit Enter.

This commit adds a command-line switch (-1) to the code that implements this
behavior.

Commit:
Do not select initial query text

When we supply an initial query, due to the way macOS's text fields work, this
initial query is selected. So, any subsequent text we type ends up replacing the
initial query rather than appending to it.

This commit makes sure that the above does not happen. When an initial query is
present, any subsequent keypresses now have the effect of appending to said
query rather than replacing it.

Commit:
Add option to match words instead of characters

Previously, choose could only match each element against the query character by
character. That is, if all characters in the query appeared in the element (in
order), the element was considered a match. Even if the characters didn't appear
in sequence.

For example, the row "banana" would match the query "bn" because the string
"banana" contains the string "bn" as a sub-sequence -- even though "b" and "n"
don't appear as consecutive characters anywhere in banana.

This commit adds an option (-W) to only consider rows as matching if the words
in the query appear in them as *continuous* sub-strings (NOT subsequences) -- in
order.

For example, with this option enabled, "banana" would match "an" but not "bn".
And "the quick brown fox" would match "qui fox" but not "qi fx" or "fox qui".

Commit:
Add option to skip sorting matches

Previously, choose would assign a score to each matching row, and present these
rows sorted by their score.

Now, I've added an option (-S) that when enabled, skips this scoring and
sorting. So, matches will simply be presented in whatever order they appeared
in the input.

Commit:
Support case-sensitive and smart-case matching

At the moment, choose does not allow us to match against queries
case-sensitively. It also does not allow smart-case (ie, case-sensitive when the
query contains uppercase characters, but case-insensitive if the query is
all-lowercase).

With this commit, we support these matching options. Specifically, I added a
command-line switch (-C) that takes 1 argument:

    - '-C i' = case-insensitive (the default),
    - '-C I' = case-sensitive, and
    - '-C s' = smart-case.

Also, I refactored some of the code for performing searching/matching. I think I
made the code a little easier to read -- especially around the parts that handle
matching from the beginning vs matching from the end. Also, I think I may have
improved performance a little: the way the code works, the same query is run
against all our choices. So, if some operation needs to be done on the query
(like splitting it into words/tokens), it's enough if we do this operation once,
and re-use the result for all our choices. That's what I'm doing in this commit.

Commit:
Merge remote-tracking branch 'origin/master'

In the time that I've been making my changes, it looks like a new commit has
been pushed to the repo -- adding functionality for choose to be used as a
password input prompt.

I have now merged my code with this commit, and confirmed that both my features
and the password feature work as intended.

chipsenkbeil commented:
Thanks for this! Letting you know that I've seen the PR, but it's going to take me a little time to review. 

Everything you've listed out as features & fixes make sense and I have no problem including them, so no concerns with the added functionality. Thanks for putting this together!

kvaadit commented:
Absolutely. Please let me know if you need any clarifications about the code I added. And many thanks for taking the time and trouble to develop/maintain such a nifty tool!

chipsenkbeil commented on SDAppDelegate.m:
```diff
@@ -790,8 +942,8 @@ - (NSArray*) getInputItems {
 #else
 
     NSFileHandle* stdinHandle = [NSFileHandle fileHandleWithStandardInput];
-    NSData* inputData = Password ? nil : [stdinHandle readDataToEndOfFile];
```
What's the reason for removing the password ternary here?

kvaadit replied:
Thanks for catching this!

Sorry, I think I missed it while resolving merge conflicts. Fixed now in commit 63cf82c16356bdaa2106fb9f6a7ed04360e35329. I also built the code and tested it. It's ready to merge I think.

chipsenkbeil reviewed (commented):
Seems good to me! Left one comment for you to check out.

Commit:
Apply input handling code missed out in last merge

While resolving conflicts during the last merge, I missed a line that processed
stdin differently depending on whether the Password option was on/off.

This commit fixes that.

Thanks to @chipsenkbeil for catching this!

chipsenkbeil commented:
Thanks for the well-written PR description and code, @kvaadit 😄 I wish that I had more contributors like you at this calibre because it would certainly make my life and the code much higher quality!

Merged and I'll publish a new release this weekend.

kvaadit commented:
Very kind of you to say, thanks @chipsenkbeil!

Looking forward to the new release.
